File: iosApp/iosApp/Modules/Networks/NetworksWireframe.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

final class DefaultNetworksWireframe {
    private weak var parent: UIViewController?
    private let alertWireframeFactory: AlertWireframeFactory
    private let networkSettingsWireframeFactory: NetworkSettingsWireframeFactory
    private let networksService: NetworksService

    private weak var vc: UIViewController?

    init(
        parent: UIViewController?,
        networksService: NetworksService,
        alertWireframeFactory: AlertWireframeFactory,
        networkSettingsWireframeFactory: NetworkSettingsWireframeFactory
    ) {
        self.parent = parent
        self.networksService = networksService
        self.alertWireframeFactory = alertWireframeFactory
        self.networkSettingsWireframeFactory = networkSettingsWireframeFactory
    }
}

extension DefaultNetworksWireframe: NetworksWireframe {

    func present() {
        let vc = wireUp()
        self.vc = vc
        parent?.asEdgeCardsController?.setTopCard(vc: vc) ?? parent?.show(vc, sender: self)
        if let parent = self.parent as? EdgeCardsController {
            parent.setTopCard(vc: vc)
        } else {
            parent?.show(vc, sender: self)
        }
    }

    func navigate(destination___________ destination: NetworksWireframeDestination) {
        if destination is NetworksWireframeDestination.Dashboard {
            parent?.asEdgeCardsController?.setDisplayMode(.master, animated: true)
        }
        if let input = destination as? NetworksWireframeDestination.EditNetwork {
            alertWireframeFactory.make(
                parent,
                context: .underConstructionAlert()
             )
 //             networkSettingsWireframeFactory.make(
 //                 vc?.asNavVc?.topVc ?? vc,
 //                 network: input.network
 //             ).present()
        }
    }
}

private extension DefaultNetworksWireframe {

    func wireUp() -> UIViewController {
        let interactor = DefaultNetworksInteractor(networksService: networksService)
        let vc: NetworksViewController = UIStoryboard(.networks).instantiate()
        let presenter = DefaultNetworksPresenter(
            view: WeakRef(referred: vc),
            wireframe: self,
            interactor: interactor
        )
        vc.presenter = presenter
        return NavigationController(rootViewController: vc)
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/networks/NetworksWireframe.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.networks

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.extensions.navigationFragment
import com.sonsofcrypto.web3wallet.android.common.ui.navigationFragment
import com.sonsofcrypto.web3walletcore.modules.networks.DefaultNetworksInteractor
import com.sonsofcrypto.web3walletcore.modules.networks.DefaultNetworksPresenter
import com.sonsofcrypto.web3walletcore.modules.networks.NetworksWireframe
import com.sonsofcrypto.web3walletcore.modules.networks.NetworksWireframeDestination

class DefaultNetworksWireframe(
    private val parent: WeakRef<Fragment>?,
    private val networksService: NetworksService,
): NetworksWireframe {

    override fun present() {
        val fragment = wireUp()
        parent?.navigationFragment?.push(fragment, animated = true)
    }

    override fun navigate(destination: NetworksWireframeDestination) {
        when (destination) {
            is NetworksWireframeDestination.Dashboard -> {
                // TODO: Navigate to dashboard
                println("Implement navigation to $destination")
            }
            is NetworksWireframeDestination.EditNetwork -> {
                // TODO: Navigate to EditNetwork
                println("Implement navigation to $destination")
            }
        }
    }

    private fun wireUp(): Fragment {
        val view = NetworksFragment()
        val interactor = DefaultNetworksInteractor(
            networksService,
        )
        val presenter = DefaultNetworksPresenter(
            WeakRef(view),
            this,
            interactor,
        )
        view.presenter = presenter
        return view
    }
}